File: apps/web/src/lib/issue-graph.ts
// EXP-980: the `blocks` graph behind the list badge, the mini-graph overlay
// and the blocked-start dialog.
//
// ONE pure rule, mirrored ×4 (iOS `ExpCore/Sources/Domain/IssueGraph.swift`,
// Android `domain/IssueGraph.kt`, desktop `domain::issue_graph`) and locked by
// the contract fixture `domain-contract/fixtures/issue-graph.json` — same
// cases, same test names.
//
// A canonical `blocks` row is `issue_id` BLOCKS `related_issue_id` (EXP-736).
// An edge counts only while BOTH ends are synced and OPEN (anchor status not
// done / cancelled / duplicate): a finished blocker is in nobody's way, and a
// finished issue is blocked by nothing. A SUBJECT is always kept, open or not.

const CLOSED_ANCHORS = new Set<string>([`done`, `cancelled`, `duplicate`])

/** The most nodes one graph draws; the rest is cut and `truncated` says so. */
export const ISSUE_GRAPH_MAX_NODES = 60

export interface GraphRelation {
  type: string
  issueId: string
  relatedIssueId: string
}

export interface GraphIssue {
  id: string
  identifier: string
  /** The dual-written ANCHOR enum (`issues.status`). */
  status: string
}

export interface BlockCounts {
  /** Open issues that block this one. */
  blockedBy: number
  /** Open issues this one blocks. */
  blocking: number
}

export interface IssueGraphNode {
  id: string
  /** The column: 0 = blocked by nothing in the graph; every edge points to a
   *  higher wave (cycle edges aside). */
  wave: number
  /** The row inside the wave, by identifier. */
  lane: number
  subject: boolean
}

export interface IssueGraphEdge {
  /** The blocker. */
  from: string
  /** The blocked issue. */
  to: string
  /** Part of a blocking cycle: drawn red, and nothing on it can start. */
  cycle: boolean
}

export interface IssueGraph {
  nodes: IssueGraphNode[]
  edges: IssueGraphEdge[]
  hasCycle: boolean
  truncated: boolean
}

/** Under a graph the node cap cut. Byte-identical ×4. */
export const ISSUE_GRAPH_TRUNCATED_NOTE = `Showing the nearest ${ISSUE_GRAPH_MAX_NODES} issues.`
/** Under a graph that holds a cycle. Byte-identical ×4. */
export const ISSUE_GRAPH_CYCLE_NOTE = `Red issues block each other in a cycle.`

/** The badge's accessible label: `Blocked by 2`, `Blocking 1` or
 *  `Blocked by 2, blocking 1`. Byte-identical ×4. */
export function blocksBadgeLabel(counts: BlockCounts): string {
  const parts: string[] = []
  if (counts.blockedBy > 0) parts.push(`Blocked by ${counts.blockedBy}`)
  if (counts.blocking > 0) {
    parts.push(`${parts.length > 0 ? `blocking` : `Blocking`} ${counts.blocking}`)
  }
  return parts.join(`, `)
}

const byText = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0)

function openEdges(
  relations: readonly GraphRelation[],
  issues: readonly GraphIssue[],
  keep: ReadonlySet<string>
): Array<[string, string]> {
  const byId = new Map(issues.map((issue) => [issue.id, issue]))
  const usable = (id: string) => {
    const issue = byId.get(id)
    return issue !== undefined && (keep.has(id) || !CLOSED_ANCHORS.has(issue.status))
  }
  const seen = new Set<string>()
  const edges: Array<[string, string]> = []
  for (const relation of relations) {
    if (relation.type !== `blocks`) continue
    if (relation.issueId === relation.relatedIssueId) continue
    if (!usable(relation.issueId) || !usable(relation.relatedIssueId)) continue
    const key = `${relation.issueId}\n${relation.relatedIssueId}`
    if (seen.has(key)) continue
    seen.add(key)
    edges.push([relation.issueId, relation.relatedIssueId])
  }
  return edges
}

/** EXP-998: every open `blocks` edge, blocker → blocked, deduplicated — what
 *  the list's rail draws (`lib/issue-rail.ts`). */
export function openBlockEdges(
  relations: readonly GraphRelation[],
  issues: readonly GraphIssue[]
): Array<{ from: string; to: string }> {
  return openEdges(relations, issues, new Set()).map(([from, to]) => ({ from, to }))
}

/** The badge numbers of every issue that has any; an issue with neither count
 *  is absent. */
export function blockCounts(
  relations: readonly GraphRelation[],
  issues: readonly GraphIssue[]
): Map<string, BlockCounts> {
  const counts = new Map<string, BlockCounts>()
  const at = (id: string) => {
    let entry = counts.get(id)
    if (!entry) {
      entry = { blockedBy: 0, blocking: 0 }
      counts.set(id, entry)
    }
    return entry
  }
  for (const [from, to] of openEdges(relations, issues, new Set())) {
    at(from).blocking += 1
    at(to).blockedBy += 1
  }
  return counts
}

/**
 * The open issues that block any of `ids` from OUTSIDE the set, by identifier:
 * what a batch start has to ask about (a blocker picked into the same batch is
 * not in its way).
 */
export function openBlockersOfSet<T extends GraphIssue>(
  ids: readonly string[],
  relations: readonly GraphRelation[],
  issues: readonly T[]
): T[] {
  const picked = new Set(ids)
  const byId = new Map(issues.map((issue) => [issue.id, issue]))
  const out = new Map<string, T>()
  for (const [from, to] of openEdges(relations, issues, new Set())) {
    if (!picked.has(to) || picked.has(from)) continue
    const blocker = byId.get(from)
    if (blocker) out.set(from, blocker)
  }
  return [...out.values()].sort((a, b) => byText(a.identifier, b.identifier))
}

/**
 * The graph around `subjectIds`: the subjects, everything that transitively
 * blocks them and everything they transitively block, with every open edge
 * among those nodes.
 *
 * Layout: an edge is a CYCLE edge when its blocker is reachable from its
 * blocked end; `wave` is the longest path over the remaining (acyclic) edges,
 * `lane` the identifier order inside a wave. Nodes come back by (wave, lane),
 * edges by (from, to) identifier.
 */
export function blockGraph(
  subjectIds: readonly string[],
  relations: readonly GraphRelation[],
  issues: readonly GraphIssue[]
): IssueGraph {
  const byId = new Map(issues.map((issue) => [issue.id, issue]))
  const identifierOf = (id: string) => byId.get(id)?.identifier ?? id
  const byIdentifier = (a: string, b: string) =>
    byText(identifierOf(a), identifierOf(b)) || byText(a, b)

  const subjects = [...new Set(subjectIds)].filter((id) => byId.has(id)).sort(byIdentifier)
  const subjectSet = new Set(subjects)
  const all = openEdges(relations, issues, subjectSet)
  const blockersOf = new Map<string, string[]>()
  const blockedBy = new Map<string, string[]>()
  for (const [from, to] of all) {
    blockersOf.set(to, [...(blockersOf.get(to) ?? []), from])
    blockedBy.set(from, [...(blockedBy.get(from) ?? []), to])
  }

  // The closure, blockers first and then the blocked side, level by level and
  // in identifier order, so the node cap cuts the same nodes everywhere.
  const picked = new Set<string>()
  let truncated = false
  const admit = (id: string) => {
    if (picked.has(id)) return false
    if (picked.size >= ISSUE_GRAPH_MAX_NODES) {
      truncated = true
      return false
    }
    picked.add(id)
    return true
  }
  for (const id of subjects) admit(id)
  for (const next of [blockersOf, blockedBy]) {
    let frontier = subjects
    while (frontier.length > 0) {
      const found = new Set<string>()
      for (const id of frontier) {
        for (const other of next.get(id) ?? []) {
          if (!picked.has(other)) found.add(other)
        }
      }
      frontier = [...found].sort(byIdentifier).filter(admit)
    }
  }

  const inside = all.filter(([from, to]) => picked.has(from) && picked.has(to))
  const out = new Map<string, string[]>()
  for (const [from, to] of inside) out.set(from, [...(out.get(from) ?? []), to])
  const reaches = (start: string, goal: string) => {
    const seen = new Set<string>([start])
    const stack = [start]
    while (stack.length > 0) {
      const id = stack.pop() as string
      if (id === goal) return true
      for (const other of out.get(id) ?? []) {
        if (!seen.has(other)) {
          seen.add(other)
          stack.push(other)
        }
      }
    }
    return false
  }

  const edges: IssueGraphEdge[] = inside
    .map(([from, to]) => ({ from, to, cycle: reaches(to, from) }))
    .sort((a, b) => byIdentifier(a.from, b.from) || byIdentifier(a.to, b.to))

  // Longest path over the acyclic edges (Kahn).
  const wave = new Map<string, number>([...picked].map((id) => [id, 0]))
  const pending = new Map<string, number>([...picked].map((id) => [id, 0]))
  const forward = new Map<string, string[]>()
  for (const edge of edges) {
    if (edge.cycle) continue
    pending.set(edge.to, (pending.get(edge.to) ?? 0) + 1)
    forward.set(edge.from, [...(forward.get(edge.from) ?? []), edge.to])
  }
  const ready = [...picked].filter((id) => pending.get(id) === 0)
  while (ready.length > 0) {
    const id = ready.pop() as string
    for (const other of forward.get(id) ?? []) {
      wave.set(other, Math.max(wave.get(other) ?? 0, (wave.get(id) ?? 0) + 1))
      const left = (pending.get(other) ?? 0) - 1
      pending.set(other, left)
      if (left === 0) ready.push(other)
    }
  }

  const ordered = [...picked].sort(
    (a, b) => (wave.get(a) ?? 0) - (wave.get(b) ?? 0) || byIdentifier(a, b)
  )
  const laneAt = new Map<number, number>()
  const nodes: IssueGraphNode[] = ordered.map((id) => {
    const w = wave.get(id) ?? 0
    const lane = laneAt.get(w) ?? 0
    laneAt.set(w, lane + 1)
    return { id, wave: w, lane, subject: subjectSet.has(id) }
  })

  return { nodes, edges, hasCycle: edges.some((edge) => edge.cycle), truncated }
}

// EXP-1057: THE mini-graph's geometry, identical ×4 and locked by
// `domain-contract/fixtures/issue-graph-geometry.json` (desktop
// `domain::issue_graph::geometry`, iOS `IssueGraph.Geometry`, Android
// `IssueGraph.Geometry`). The web draws it through `@exp/ui` `WaveGraph`,
// whose edge rule is the one below; the grid sits `inset` inside its scroll
// box so the rings are never clipped.
export const ISSUE_GRAPH_GEOMETRY = {
  nodeWidth: 176,
  nodeHeight: 28,
  waveGap: 40,
  laneGap: 8,
  inset: 4,
  maxViewWidth: 520,
  maxViewHeight: 320,
  edgeStroke: 1.25,
  ringWidth: 1,
  nodeRadius: 6,
  railGutter: 8,
  railNodeWidth: 24,
  railDot: 10,
  railDotRing: 2,
} as const

export interface GraphPoint {
  x: number
  y: number
}

/** A node box's top-left inside the grid. */
export function issueGraphOrigin(wave: number, lane: number): GraphPoint {
  const g = ISSUE_GRAPH_GEOMETRY
  return {
    x: g.inset + wave * (g.nodeWidth + g.waveGap),
    y: g.inset + lane * (g.nodeHeight + g.laneGap),
  }
}

/** The grid's natural size for `waves` × `lanes` (insets included) and the
 *  viewport it shows before scrolling. */
export function issueGraphSize(
  waves: number,
  lanes: number
): { width: number; height: number; viewWidth: number; viewHeight: number } {
  const g = ISSUE_GRAPH_GEOMETRY
  if (waves <= 0 || lanes <= 0) {
    return { width: 0, height: 0, viewWidth: 0, viewHeight: 0 }
  }
  const width = 2 * g.inset + waves * (g.nodeWidth + g.waveGap) - g.waveGap
  const height = 2 * g.inset + lanes * (g.nodeHeight + g.laneGap) - g.laneGap
  return {
    width,
    height,
    viewWidth: Math.min(width, g.maxViewWidth),
    viewHeight: Math.min(height, g.maxViewHeight),
  }
}

/** One edge as a cubic: blocker's right-middle → blocked box's left-middle. */
export function issueGraphEdge(
  from: { wave: number; lane: number },
  to: { wave: number; lane: number }
): { start: GraphPoint; control1: GraphPoint; control2: GraphPoint; end: GraphPoint } {
  const g = ISSUE_GRAPH_GEOMETRY
  const a = issueGraphOrigin(from.wave, from.lane)
  const b = issueGraphOrigin(to.wave, to.lane)
  const start = { x: a.x + g.nodeWidth, y: a.y + g.nodeHeight / 2 }
  const end = { x: b.x, y: b.y + g.nodeHeight / 2 }
  const bend =
    end.x > start.x
      ? (end.x - start.x) / 2
      : Math.max(g.waveGap / 2, Math.abs(end.x - start.x) / 2)
  return {
    start,
    control1: { x: start.x + bend, y: start.y },
    control2: { x: end.x - bend, y: end.y },
    end,
  }
}

File: apps/ios/ExpCore/Sources/Domain/IssueGraph.swift
import Foundation

/// EXP-980 — the `blocks` graph behind the list badge, the mini-graph overlay
/// and the blocked-start dialog.
///
/// ONE pure rule, mirrored ×4 (web `lib/issue-graph.ts`, Android
/// `domain/IssueGraph.kt`, desktop `domain::issue_graph`) and locked by the
/// contract fixture `domain-contract/fixtures/issue-graph.json` — same cases,
/// same test names.
///
/// A canonical `blocks` row is `issueId` BLOCKS `relatedIssueId` (EXP-736). An
/// edge counts only while BOTH ends are synced and OPEN (anchor status not
/// done / cancelled / duplicate): a finished blocker is in nobody's way, and a
/// finished issue is blocked by nothing. A SUBJECT is always kept, open or not.
public enum IssueGraph {

    /// The most nodes one graph draws; the rest is cut and `truncated` says so.
    public static let maxNodes = 60

    /// Under a graph the node cap cut. Byte-identical ×4.
    public static let truncatedNote = "Showing the nearest \(maxNodes) issues."
    /// Under a graph that holds a cycle. Byte-identical ×4.
    public static let cycleNote = "Red issues block each other in a cycle."

    /// One row's badge numbers.
    public struct BlockCounts: Equatable, Sendable {
        /// Open issues that block this one.
        public let blockedBy: Int
        /// Open issues this one blocks.
        public let blocking: Int

        public init(blockedBy: Int, blocking: Int) {
            self.blockedBy = blockedBy
            self.blocking = blocking
        }
    }

    public struct Node: Equatable, Sendable {
        public let id: String
        /// The column: 0 = blocked by nothing in the graph; every edge points
        /// to a higher wave (cycle edges aside).
        public let wave: Int
        /// The row inside the wave, by identifier.
        public let lane: Int
        public let subject: Bool

        public init(id: String, wave: Int, lane: Int, subject: Bool) {
            self.id = id
            self.wave = wave
            self.lane = lane
            self.subject = subject
        }
    }

    public struct Edge: Equatable, Sendable {
        /// The blocker.
        public let from: String
        /// The blocked issue.
        public let to: String
        /// Part of a blocking cycle: drawn red, and nothing on it can start.
        public let cycle: Bool

        public init(from: String, to: String, cycle: Bool) {
            self.from = from
            self.to = to
            self.cycle = cycle
        }
    }

    public struct Graph: Equatable, Sendable {
        public let nodes: [Node]
        public let edges: [Edge]
        public let hasCycle: Bool
        public let truncated: Bool

        public init(nodes: [Node], edges: [Edge], hasCycle: Bool, truncated: Bool) {
            self.nodes = nodes
            self.edges = edges
            self.hasCycle = hasCycle
            self.truncated = truncated
        }

        public var isEmpty: Bool { nodes.isEmpty }
    }

    /// The badge's accessible label: `Blocked by 2`, `Blocking 1` or
    /// `Blocked by 2, blocking 1`. Byte-identical ×4.
    public static func blocksBadgeLabel(_ counts: BlockCounts) -> String {
        var parts: [String] = []
        if counts.blockedBy > 0 { parts.append("Blocked by \(counts.blockedBy)") }
        if counts.blocking > 0 {
            parts.append("\(parts.isEmpty ? "Blocking" : "blocking") \(counts.blocking)")
        }
        return parts.joined(separator: ", ")
    }

    /// The badge numbers of every issue that has any; an issue with neither
    /// count is absent.
    public static func blockCounts(
        relations: [IssueRelationEntity],
        issues: [IssueEntity]
    ) -> [String: BlockCounts] {
        var blockedBy: [String: Int] = [:]
        var blocking: [String: Int] = [:]
        for (from, to) in openEdges(relations, issues, keep: []) {
            blocking[from, default: 0] += 1
            blockedBy[to, default: 0] += 1
        }
        var counts: [String: BlockCounts] = [:]
        for id in Set(blockedBy.keys).union(blocking.keys) {
            counts[id] = BlockCounts(
                blockedBy: blockedBy[id] ?? 0, blocking: blocking[id] ?? 0
            )
        }
        return counts
    }

    /// The open issues that block any of `ids` from OUTSIDE the set, by
    /// identifier: what a batch start has to ask about (a blocker picked into
    /// the same batch is not in its way).
    public static func openBlockersOfSet(
        ids: [String],
        relations: [IssueRelationEntity],
        issues: [IssueEntity]
    ) -> [IssueEntity] {
        let picked = Set(ids)
        let byId = Dictionary(issues.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        var seen = Set<String>()
        var out: [IssueEntity] = []
        for (from, to) in openEdges(relations, issues, keep: []) {
            guard picked.contains(to), !picked.contains(from) else { continue }
            guard !seen.contains(from), let blocker = byId[from] else { continue }
            seen.insert(from)
            out.append(blocker)
        }
        return out.sorted { sortKey($0) < sortKey($1) }
    }

    /// The graph around `subjectIds`: the subjects, everything that
    /// transitively blocks them and everything they transitively block, with
    /// every open edge among those nodes.
    ///
    /// Layout: an edge is a CYCLE edge when its blocker is reachable from its
    /// blocked end; `wave` is the longest path over the remaining (acyclic)
    /// edges, `lane` the identifier order inside a wave. Nodes come back by
    /// (wave, lane), edges by (from, to) identifier.
    public static func blockGraph(
        subjectIds: [String],
        relations: [IssueRelationEntity],
        issues: [IssueEntity]
    ) -> Graph {
        let byId = Dictionary(issues.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        // The identifier is the ordering key everywhere; an id stands in for a
        // row that never stamped one, so the order stays total.
        func key(_ id: String) -> (String, String) { (byId[id]?.identifier ?? id, id) }

        let subjects = Set(subjectIds).filter { byId[$0] != nil }
            .sorted { key($0) < key($1) }
        let subjectSet = Set(subjects)
        let all = openEdges(relations, issues, keep: subjectSet)
        var blockersOf: [String: [String]] = [:]
        var blockedBy: [String: [String]] = [:]
        for (from, to) in all {
            blockersOf[to, default: []].append(from)
            blockedBy[from, default: []].append(to)
        }

        // The closure, blockers first and then the blocked side, level by
        // level and in identifier order, so the node cap cuts the same nodes
        // everywhere.
        var picked = Set<String>()
        var truncated = false
        func admit(_ id: String) -> Bool {
            if picked.contains(id) { return false }
            if picked.count >= maxNodes {
                truncated = true
                return false
            }
            picked.insert(id)
            return true
        }
        for id in subjects { _ = admit(id) }
        for next in [blockersOf, blockedBy] {
            var frontier = subjects
            while !frontier.isEmpty {
                var found = Set<String>()
                for id in frontier {
                    for other in next[id] ?? [] where !picked.contains(other) {
                        found.insert(other)
                    }
                }
                frontier = found.sorted { key($0) < key($1) }.filter { admit($0) }
            }
        }

        let inside = all.filter { picked.contains($0.0) && picked.contains($0.1) }
        var out: [String: [String]] = [:]
        for (from, to) in inside { out[from, default: []].append(to) }
        func reaches(_ start: String, _ goal: String) -> Bool {
            var seen: Set<String> = [start]
            var stack = [start]
            while let id = stack.popLast() {
                if id == goal { return true }
                for other in out[id] ?? [] where !seen.contains(other) {
                    seen.insert(other)
                    stack.append(other)
                }
            }
            return false
        }

        let edges: [Edge] = inside
            .map { Edge(from: $0.0, to: $0.1, cycle: reaches($0.1, $0.0)) }
            .sorted { a, b in
                if key(a.from) != key(b.from) { return key(a.from) < key(b.from) }
                return key(a.to) < key(b.to)
            }

        // Longest path over the acyclic edges (Kahn).
        var wave: [String: Int] = [:]
        var pending: [String: Int] = [:]
        for id in picked {
            wave[id] = 0
            pending[id] = 0
        }
        var forward: [String: [String]] = [:]
        for edge in edges where !edge.cycle {
            pending[edge.to, default: 0] += 1
            forward[edge.from, default: []].append(edge.to)
        }
        var ready = Array(picked).filter { pending[$0] == 0 }
        while let id = ready.popLast() {
            for other in forward[id] ?? [] {
                wave[other] = max(wave[other] ?? 0, (wave[id] ?? 0) + 1)
                let left = (pending[other] ?? 0) - 1
                pending[other] = left
                if left == 0 { ready.append(other) }
            }
        }

        let ordered = picked.sorted { a, b in
            let (wa, wb) = (wave[a] ?? 0, wave[b] ?? 0)
            if wa != wb { return wa < wb }
            return key(a) < key(b)
        }
        var laneAt: [Int: Int] = [:]
        let nodes: [Node] = ordered.map { id in
            let w = wave[id] ?? 0
            let lane = laneAt[w] ?? 0
            laneAt[w] = lane + 1
            return Node(id: id, wave: w, lane: lane, subject: subjectSet.contains(id))
        }

        return Graph(
            nodes: nodes,
            edges: edges,
            hasCycle: edges.contains { $0.cycle },
            truncated: truncated
        )
    }

    // MARK: - Geometry

    /// EXP-1057: THE mini-graph's geometry, identical ×4 and locked by
    /// `domain-contract/fixtures/issue-graph-geometry.json` (web
    /// `lib/issue-graph.ts` `ISSUE_GRAPH_GEOMETRY`, desktop
    /// `domain::issue_graph::geometry`, Android `IssueGraph.Geometry`). Points:
    /// web px = desktop px = iOS pt = Android dp. The grid sits `inset` inside
    /// its scroll box so the rings are never clipped.
    public enum Geometry {
        public static let nodeWidth: Double = 176
        public static let nodeHeight: Double = 28
        public static let waveGap: Double = 40
        public static let laneGap: Double = 8
        public static let inset: Double = 4
        public static let maxViewWidth: Double = 520
        public static let maxViewHeight: Double = 320
        public static let edgeStroke: Double = 1.25
        public static let ringWidth: Double = 1
        public static let nodeRadius: Double = 6
        public static let railGutter: Double = 8
        public static let railNodeWidth: Double = 24
        public static let railDot: Double = 10
        public static let railDotRing: Double = 2

        public struct Point: Equatable, Sendable {
            public let x: Double
            public let y: Double

            public init(x: Double, y: Double) {
                self.x = x
                self.y = y
            }
        }

        /// The grid's natural size (insets included) and the viewport it shows
        /// before scrolling.
        public struct Size: Equatable, Sendable {
            public let width: Double
            public let height: Double
            public let viewWidth: Double
            public let viewHeight: Double

            public init(width: Double, height: Double, viewWidth: Double, viewHeight: Double) {
                self.width = width
                self.height = height
                self.viewWidth = viewWidth
                self.viewHeight = viewHeight
            }
        }

        /// One edge as a cubic: blocker's right-middle → blocked box's left-middle.
        public struct EdgeCurve: Equatable, Sendable {
            public let start: Point
            public let control1: Point
            public let control2: Point
            public let end: Point

            public init(start: Point, control1: Point, control2: Point, end: Point) {
                self.start = start
                self.control1 = control1
                self.control2 = control2
                self.end = end
            }
        }

        /// A node box's top-left inside the grid.
        public static func origin(wave: Int, lane: Int) -> Point {
            Point(
                x: inset + Double(wave) * (nodeWidth + waveGap),
                y: inset + Double(lane) * (nodeHeight + laneGap)
            )
        }

        /// The grid for `waves` × `lanes`; nothing at all when either is 0.
        public static func size(waves: Int, lanes: Int) -> Size {
            guard waves > 0, lanes > 0 else {
                return Size(width: 0, height: 0, viewWidth: 0, viewHeight: 0)
            }
            let width = 2 * inset + Double(waves) * (nodeWidth + waveGap) - waveGap
            let height = 2 * inset + Double(lanes) * (nodeHeight + laneGap) - laneGap
            return Size(
                width: width,
                height: height,
                viewWidth: min(width, maxViewWidth),
                viewHeight: min(height, maxViewHeight)
            )
        }

        /// The grid a graph needs: its highest wave and lane, plus one.
        public static func size(of graph: Graph) -> Size {
            size(
                waves: (graph.nodes.map(\.wave).max() ?? -1) + 1,
                lanes: (graph.nodes.map(\.lane).max() ?? -1) + 1
            )
        }

        /// Forward, the curve bends on the gap's middle; a backward (cycle)
        /// edge bows by max(waveGap / 2, |dx| / 2).
        public static func edge(
            from: (wave: Int, lane: Int),
            to: (wave: Int, lane: Int)
        ) -> EdgeCurve {
            let a = origin(wave: from.wave, lane: from.lane)
            let b = origin(wave: to.wave, lane: to.lane)
            let start = Point(x: a.x + nodeWidth, y: a.y + nodeHeight / 2)
            let end = Point(x: b.x, y: b.y + nodeHeight / 2)
            let bend = end.x > start.x
                ? (end.x - start.x) / 2
                : max(waveGap / 2, abs(end.x - start.x) / 2)
            return EdgeCurve(
                start: start,
                control1: Point(x: start.x + bend, y: start.y),
                control2: Point(x: end.x - bend, y: end.y),
                end: end
            )
        }
    }

    // MARK: - Pieces

    /// The `blocks` edges both of whose ends are synced and open, deduped and
    /// in row order. `keep` are the subjects, which stay whatever their status.
    private static func openEdges(
        _ relations: [IssueRelationEntity],
        _ issues: [IssueEntity],
        keep: Set<String>
    ) -> [(String, String)] {
        let byId = Dictionary(issues.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        func usable(_ id: String) -> Bool {
            guard let issue = byId[id] else { return false }
            return keep.contains(id) || !isFinished(issue)
        }
        var seen = Set<String>()
        var edges: [(String, String)] = []
        for relation in relations
        where relation.type == IssueRelationType.blocks.rawValue {
            if relation.issueId == relation.relatedIssueId { continue }
            guard usable(relation.issueId), usable(relation.relatedIssueId) else { continue }
            let key = "\(relation.issueId)\n\(relation.relatedIssueId)"
            if seen.contains(key) { continue }
            seen.insert(key)
            edges.append((relation.issueId, relation.relatedIssueId))
        }
        return edges
    }

    /// The terminal anchors: an issue in one of them blocks nothing, and is
    /// blocked by nothing.
    private static func isFinished(_ issue: IssueEntity) -> Bool {
        switch IssueStatus.from(issue.status) {
        case .done, .cancelled, .duplicate: return true
        case .backlog, .inProgress, .inReview: return false
        }
    }

    /// Identifier order, id as the deterministic tie-break.
    private static func sortKey(_ issue: IssueEntity) -> (String, String) {
        (issue.identifier ?? issue.id, issue.id)
    }
}
